### PR TITLE
DT-462: Add RFC for Reek configuration

### PR DESCRIPTION
**Pull request by arielj**
Read the .md file included in the DIFF for details.

You can see the parsed version of the file at https://github.com/fastruby/RFCs/blob/DT-462-shared-default-reek-configuration/2025-06-02-default-reek-configuration.md

### CODE REVIEW

File: 2025-06-02-default-reek-configuration.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+This RFC is a proposal of a default Reek configuration file that we would use in established projects to reduce the number of warnings and the need of exceptions that we see with the [default configuration](https://github.com/troessner/reek/blob/a8649370ced77dd0e798b00dc97ff916f68bb002/docs/defaults.reek.yml), and to also define some guideline or expectation on how we can decide when to add exceptions and modifications for different needs.
+
+### Default .reek configuration file
+
+This file is meant to be added in the root of the project as `.reek.yml` to handle common cases of Reek warnings that will want either disabled or reconfigured based on what we've seen in many projects.
+
+It's a combination of the [official `rails-friendly` configuration recommended in the Reek's README](https://github.com/troessner/reek?tab=readme-ov-file#working-with-rails) and some changes to the default rules.
+
+```yml
+detectors:
```
Comment (fbuys) [inline 2025-06-02-default-reek-configuration.md:18]: I think this is useful too.
It allows us to do `rescue e =>` and `map { |x| x.do_something }`

```
  UncommunicativeVariableName:
    accept:
     - e
     - x
```

File: 2025-06-02-default-reek-configuration.md
Comment (JuanVqz) [inline 2025-06-02-default-reek-configuration.md:1]: would it be possible to propose a new rule? if so, where is the best place to do it?

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:1]: comments here in the PR is the place for that

File: 2025-06-02-default-reek-configuration.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  IrresponsibleModule:
+    enabled: false # we don't really add module descriptions ever
+  TooManyStatements:
+    max_statements: 10 # original is 5
+  DuplicateMethodCall:
+    max_calls: 3 # original is 1
```
Comment (FionaDL) [inline 2025-06-02-default-reek-configuration.md:20]: Is this enough for tests?

File: 2025-06-02-default-reek-configuration.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+## How is this going to impact our internal teams and their day to day?
+
+In the near future we plan to adapt this configuration in these projects:
+
+- FastRuby.io
+- OmbuLabs.com
+- Points
+
+Other project will not adapt this configuration. We expect Ruby and Rails projects to reach certain maturity before we adapt these standards.
+
```
Comment (FionaDL) [inline 2025-06-02-default-reek-configuration.md:86]: I'm curious about this part, while I understand not wanting to slow down development in these smaller, newer projects...at what point do you add this, and how much time does it then take to go back and make sure everything is agreeing with reek standards? Where will the line fall for us?

File: 2025-06-02-default-reek-configuration.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    enabled: true # default is false, we are still disabling it for controllers and models below
+  UtilityFunction:
+    public_methods_only: true # original is false
+
+directories: # reek's recommendation for Rails applications unless a comment is added
+  "app/controllers":
+    NestedIterators:
+      max_allowed_nesting: 2
+    UnusedPrivateMethod:
+      enabled: false
```
Comment (mateusdeap) [inline 2025-06-02-default-reek-configuration.md]: I'm wondering if it wouldn't be better to have it be true. Even in controllers and models, if it's unused, wouldn't we want to remove them?

File: 2025-06-02-default-reek-configuration.md (hunk 3 above)
Comment (mateusdeap) [reply 2025-06-02-default-reek-configuration.md:86]: I feel like this is answered below, somewhat, but now I'm wondering at what point do we consider a project is not an MVP/Prototype or a one off internal tool that might be obsolete in a year...

File: 2025-06-02-default-reek-configuration.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+exclude_paths:
+  - node_modules # in case any node package has a .rb file (like the styleguide)
+  - vendor # in case gems are bundle inside the project's folder vendor folder
+```
+
+### Exceptions and :reek:\* Comments
+
+When applying this configuration, projects will still need exceptions in some cases where the code smell is not flagging code we want to change or when there's a warning on code that's in the same file but out of the scope of the work done.
+
+It is expected that the team working on a given project will keep an eye when these exceptions are used, and there should be a clear explanation to skip a code smell instead of fixing it (which is a valid thing to do, but shouldn't be done just in order to not address it).
```
Comment (mateusdeap) [inline 2025-06-02-default-reek-configuration.md:75]: I think I'd like some examples of this. I don't know of any off the top of my head, but I'm only thinking that, situationally, if a smell requires a fix that we know the client doesn't really care about or that, in our evaluation, isn't worth the time out of the project it would take to fix it, then we could skip it.

Is that it?

File: 2025-06-02-default-reek-configuration.md (hunk 1 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:18]: sounds good

File: 2025-06-02-default-reek-configuration.md (hunk 2 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:20]: what do you mean?

File: 2025-06-02-default-reek-configuration.md (hunk 4 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md]: the problem with this rule is that (at least from my tests), it's not smart enough to detect that you are referencing a private method in a callback, as in `before_action :some_private_method` on a controller or an `after_save :do_something_private` in a model

we have that enabled above in general, but we are disabling them for controllers and models

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md]: it was showing many false positives saying a private method was not used when it was used in a callback

File: 2025-06-02-default-reek-configuration.md (hunk 5 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:75]: this is not for client project, it's only for our internal projects

and example here will be that with these defaults, when we run reek in fastruby.io/ombulabs.com/points we will have reek code smells, if something reek complains about is not worth the time to fix it (maybe a method is long because it has to be long, or maybe it makes total sense to have a utility function in a file and moving it somewhere else would make it less readable), then we can add reek comments to skip it

File: 2025-06-02-default-reek-configuration.md (hunk 3 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:86]: cc @etagwerker @ABizzinotto here, I don't know if we have a clear definition of when this crosses that line (for example, in that list we are not listing UpgradeJS.com)

one thing though is that when we want to add reek to a project, it can be added with TODOs to address things gradually, it doesn't need to be something like "we add reek and now we have to fix all the smells", it's more like "we add reek, and we can address the smells gradually over time along with other changes"

File: 2025-06-02-default-reek-configuration.md (hunk 5 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:75]: another example could be a private method in some class that is not used explicitly but because it's used by some `send(...)` call with variables, in that cases reek wouldn't know it's actually used

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:75]: the main idea for this is to not encourage people to just add `:reek:` comments because they didn't feel like fixing a reek warning, if something won't be addressed there has to be a good argument for that (good as in the people working on that task/pr/story agree with)

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:75]: and this can also help use improve these defaults, if we see something that gets skipped a lot, we might be able to change these defaults based on those reasons, but we need a reason

File: 2025-06-02-default-reek-configuration.md (hunk 3 above)
Comment (FionaDL) [reply 2025-06-02-default-reek-configuration.md:86]: Won't that break the build though?

File: 2025-06-02-default-reek-configuration.md (hunk 2 above)
Comment (FionaDL) [reply 2025-06-02-default-reek-configuration.md:20]: Maybe I am wrong, but from what I remember the duplicate method calls happen often in tests when you are calling a method several times??

File: 2025-06-02-default-reek-configuration.md (hunk 1 above)
Comment (fbuys) [reply 2025-06-02-default-reek-configuration.md:18]: I wonder if we should also add `_` for those times we have a unused variable?

File: 2025-06-02-default-reek-configuration.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    max_calls: 3 # original is 1
+  NilCheck:
+    enabled: false # in some projects were this was enabled, it was a false positive being ignored
+  UnusedPrivateMethod:
+    enabled: true # default is false, we are still disabling it for controllers and models below
+  UtilityFunction:
+    public_methods_only: true # original is false
+
+directories: # reek's recommendation for Rails applications unless a comment is added
+  "app/controllers":
```
Comment (fbuys) [inline 2025-06-02-default-reek-configuration.md:40]: In the past I have found the following helpful in controllers too:
```
    FeatureEnvy:
      enabled: false
    TooManyMethods:
      enabled: false
    TooManyStatements:
      enabled: false
```

FeatureEnvy: It is easy for controller methods to interact with other parts of the application
TooManyMethods: Controllers can easily go over 15 methods
TooManyStatements: Controllers methods easily go over 5 lines

File: 2025-06-02-default-reek-configuration.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    UncommunicativeVariableName:
+      enabled: false # complains `t` is uncommunicative, but it's a common short name for `table`
+    TooManyStatements:
+      enabled: false # long tables require many statements, reek keeps complaining about them
+    DuplicateMethodCall:
+      enabled: false # complains about duplicated methods when a migration creates more than 1 table, which cannot be "fixed"
+    UtilityFunction:
+      enabled: false # complains about utility functions in generated migrations, it's ok to have utility functions in migrations
+
+exclude_paths:
```
Comment (fbuys) [inline 2025-06-02-default-reek-configuration.md:65]: `  - db/migrate` In the past I ignored the db/migrate directory too

File: 2025-06-02-default-reek-configuration.md (hunk 3 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:86]: if we add the reek ignore comments it shouldn't break the builds, it would be useful to flag the comments as something that is agreed to be ignore vs the ones we ignore to deal with them later with a TODO

File: 2025-06-02-default-reek-configuration.md (hunk 2 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:20]: I just tried calling a method multiple times in a single test and reek doesn't complain about it

I'm not sure how it works internally but it seems it won't complain during tests? we can always change this later to disable it for the test and spec directories I guess if it starts creating issues (maybe there's some edge case I don't know how to trigger)

File: 2025-06-02-default-reek-configuration.md (hunk 7 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:65]: we are already disabling many checks for the `db/migrate` directory above, do you think it's better to just ignore everything instead?

File: 2025-06-02-default-reek-configuration.md (hunk 6 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:40]: do you think those are better fully disabled or maybe add as exceptions when needed?

I'm running `reek` on fastruby with the current state of this proposed config and I don't get any TooManyMethods nor FeatureEnvy warnings for controllers, and only 2 TooManyStatements in one of the controllers (maybe those actions can be ignored as exceptions or refactored)

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:40]: maybe we can disable TooManyMethods for controllers and models too? I don't think it adds that much value (I think we are already good at not putting methods where they don't belong, and when they are in the right place it would be annoying to have reek complain just because of the number)

File: 2025-06-02-default-reek-configuration.md (hunk 1 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:18]: added these 3 accepted short names

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:18]: I wonder if we should add `k` and `v` as they are common short names for `key` and `value` when iterating over a hash

Comment (mateusdeap) [reply 2025-06-02-default-reek-configuration.md:18]: Sounds good to me

File: 2025-06-02-default-reek-configuration.md (hunk 6 above)
Comment (fbuys) [reply 2025-06-02-default-reek-configuration.md:40]: I am onboard with disabling `TooManyMethods` in controllers/models yes.
And perhaps you are right, we can disable feature envy and too many statements where needed.

File: 2025-06-02-default-reek-configuration.md (hunk 7 above)
Comment (fbuys) [reply 2025-06-02-default-reek-configuration.md:65]: I don't know if there is much value in reeking `db/migrate` file at all.
They typically fall outside of the normal application - and should not really contain business logic at all.
So I would say yes, let's disable all of the rules in this directory.

File: 2025-06-02-default-reek-configuration.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+It is expected that the team working on a given project will keep an eye when these exceptions are used, and there should be a clear explanation to skip a code smell instead of fixing it (which is a valid thing to do, but shouldn't be done just in order to not address it).
+
+## How is this going to impact our internal teams and their day to day?
+
+In the near future we plan to adapt this configuration in these projects:
+
+- FastRuby.io
+- OmbuLabs.com
+- Points
+
```
Comment (hmdros) [inline 2025-06-02-default-reek-configuration.md:84]: Not the UpgradeJS?

Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:84]: Nope, not UpgradeJS.

Even though it's a project we've had for many years, it's still a project that does not have regular updates and it's not at the level of importance of the other 3 listed here.

File: 2025-06-02-default-reek-configuration.md (hunk 1 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:18]: I updated this to include `k` and `v`

File: 2025-06-02-default-reek-configuration.md (hunk 6 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:40]: I added the change to ignore `TooManyMethods` for models and controllers

File: 2025-06-02-default-reek-configuration.md (hunk 7 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:65]: Added `db/migrate` to the list of ignored paths

File: 2025-06-02-default-reek-configuration.md (hunk 3 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:86]: I added a line clarifying that Ops will decide when a project reaches the maturity to implement this. It will be a project-specific decision.

**Review by JuanVqz — COMMENTED** [review]
Looks good to me as initial RFC :shipit:

File: 2025-06-02-default-reek-configuration.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+```yml
+detectors:
+  DuplicateMethodCall:
+    max_calls: 3 # original is 1
+  IrresponsibleModule:
+    enabled: false # we don't really add module descriptions ever
+  NilCheck:
+    enabled: false # in some projects were this was enabled, it was a false positive being ignored
+  TooManyStatements:
+    max_statements: 10 # original is 5
```
Comment (rishijain) [inline 2025-06-02-default-reek-configuration.md:26]: I think for some of the client projects, we might be easily breaking this rule as we do not have any control on the state at which we start working on the project.

Comment (rishijain) [reply 2025-06-02-default-reek-configuration.md:26]: I see that TooManyStatements is set to false for controllers and models. I think that do cover some of my concern. Although I have seen clients putting more than 10 lines of code in methods that are placed under `lib`, `config`.

**Comment by rishijain** [comment]
I think we can add these as well to the starting point:


Long Parameter List (https://github.com/troessner/reek/blob/master/docs/Long-Parameter-List.md) - This encourages better design.

TooManyInstanceVariables - Right now, we have set it to false. Although I think we should set it to true with may be a higher number of variables allowed. It signals that classes doing too much.

Data Clump (https://github.com/troessner/reek/blob/master/docs/Data-Clump.md) - It also encourages the idea of using objects and cleaner design.

File: 2025-06-02-default-reek-configuration.md (hunk 9 above)
Comment (arielj) [reply 2025-06-02-default-reek-configuration.md:26]: this will not be used on client projects, only for internal projects and only for FastRuby.io, Ombulabs.com, and Points for now

**Comment by arielj** [comment]
> I think we can add these as well to the starting point:
> 
> Long Parameter List (https://github.com/troessner/reek/blob/master/docs/Long-Parameter-List.md) - This encourages better design.
> 
> TooManyInstanceVariables - Right now, we have set it to false. Although I think we should set it to true with may be a higher number of variables allowed. It signals that classes doing too much.
> 
> Data Clump (https://github.com/troessner/reek/blob/master/docs/Data-Clump.md) - It also encourages the idea of using objects and cleaner design.

LongParameterList and DataClump are enabled by default already, and we are not disabling them

for TooManyInstanceVariables, this is also enabled by default, we are disabling it only for controllers because we usually set many instance variables to pass data to the views

**Comment by rishijain** [comment]
👍

**Review by julioalucero — APPROVED** [review]
It is nice to have this new configuration as a starting point!
